src/frontend/src/optimizer/plan_expr_visitor/strong.rs

```diff
@@ -0,0 +1,306 @@
+
+    // Returns whether the analyzed expression will definitely return null if
+    // all of a given set of input columns are null.
+    #[allow(dead_code)]
+    pub fn is_null(expr: &ExprImpl, null_columns: FixedBitSet) -> bool {
+        let strong = Strong::new(null_columns);
+        strong.is_null_visit(expr)
+    }
+
+    fn is_input_ref_null(&self, input_ref: &InputRef) -> bool {
+        self.null_columns.contains(input_ref.index())
+    }
+
+    fn is_null_visit(&self, expr: &ExprImpl) -> bool {
+        match expr {
+            ExprImpl::InputRef(input_ref) => self.is_input_ref_null(input_ref),
+            ExprImpl::Literal(literal) => literal.get_data().is_none(),
+            ExprImpl::FunctionCall(func_call) => self.is_null_function_call(func_call),
+            ExprImpl::FunctionCallWithLambda(_) => false,
+            ExprImpl::AggCall(_) => false,
+            ExprImpl::Subquery(_) => false,
+            ExprImpl::CorrelatedInputRef(_) => false,
+            ExprImpl::TableFunction(_) => false,
+            ExprImpl::WindowFunction(_) => false,
+            ExprImpl::UserDefinedFunction(_) => false,
+            ExprImpl::Parameter(_) => false,
+            ExprImpl::Now(_) => false,
+        }
+    }
+
+    fn is_null_function_call(&self, func_call: &FunctionCall) -> bool {
+        match func_call.func_type() {
+            // NOT NULL: This kind of expression is never null. No need to look at its arguments, if it has any.
+            ExprType::IsNull
+            | ExprType::IsNotNull
+            | ExprType::IsDistinctFrom
+            | ExprType::IsNotDistinctFrom
+            | ExprType::IsTrue
+            | ExprType::IsNotTrue
+            | ExprType::IsFalse
+            | ExprType::IsNotFalse => false,
+            // ANY: This kind of expression is null if and only if at least one of its arguments is null.
+            ExprType::Not
+            | ExprType::Equal
+            | ExprType::NotEqual
+            | ExprType::LessThan
+            | ExprType::LessThanOrEqual
+            | ExprType::GreaterThan
+            | ExprType::GreaterThanOrEqual
+            | ExprType::Like
+            | ExprType::Add
+            | ExprType::AddWithTimeZone
+            | ExprType::Subtract
+            | ExprType::Multiply
+            | ExprType::Modulus
+            | ExprType::Divide
+            | ExprType::Cast
+            | ExprType::Trim
+            | ExprType::Ltrim
+            | ExprType::Rtrim
+            | ExprType::Ceil
+            | ExprType::Floor
+            | ExprType::Extract
+            | ExprType::Greatest
+            | ExprType::Least => self.any_null(func_call),
+            ExprType::And | ExprType::Or | ExprType::Coalesce => self.all_null(func_call),
+            // TODO: Function like case when is important but current its structure is complicated, so we need to implement it later if necessary.
+            // Assume that any other expressions cannot be simplified.
+            _ => false,
```

Comment:
Just a quick semantic question, does return false here **semantically** represent `cannot be simplified`?
Since when we call `is_null`, we are essentially checking if the input expression will return `null`, and if `is_null` return false in this arm, and the to-be-checked expression could indeed return `null` in some cases, wouldn't this be semantically wrong / conflict especially to the caller of `is_null`?

Reply:
Yes. The `is_null` semantic is `Returns whether the analyzed expression will definitely return null if all of a given set of input columns are null`, so if this method returns false, it means we have no guarantee on anything. It doesn't mean this expr couldn't be null.

Reply:
Got it! Then what about adding the semantic description in the comment, so that we can (hopefully) use this correctly. 🤡

Reply:
We have a description right above that method.